pkg/arvo/app/permission-group-hook.hoon

```diff
@@ -0,0 +1,179 @@
+/-  *group-store, *permission-store, *permission-group-hook
+|%
++$  move  [bone card]
+::
++$  card
+  $%  [%diff [%group-update group-update]]
+      [%quit ~]
+      [%poke wire dock poke]
+      [%pull wire dock ~]
+      [%peer wire dock path]
+  ==
+::
++$  state
+  $%  [%0 state-zero]
+  ==
+::
++$  group-path  path
+::
++$  permission-path  path
+::
++$  state-zero
+  $:  relation=(map group-path (set permission-path))
+  ==
+::
++$  poke
+  $%  [%permission-action permission-action]
+  ==
+::
+--
+::
+|_  [bol=bowl:gall state]
+::
+++  this  .
+::
+++  prep
+  |=  old=(unit state)
+  ^-  (quip move _this)
+  ?~  old
+    [~ this]
+  [~ this(+<+ u.old)]
+::
+++  poke-permission-group-hook-action
+  |=  act=permission-group-hook-action
+  ^-  (quip move _this)
+  ?.  =(src.bol our.bol)
+    [~ this]
+  ?-  -.act
+      %associate
+    (handle-associate group.act permissions.act)
+  ::
+      %dissociate
+    (handle-dissociate group.act permissions.act)
+  ::
+  ==
+::
+++  handle-associate
+  |=  [group=path permissions=(set path)]
+  ^-  (quip move _this)
+  ::  TODO: remove the requirement for permission paths to exist
+  =/  perm-list  ~(tap in permissions)
+  =/  perm-length  (lent perm-list)
+  =/  checked-paths
+    %+  skim  perm-list
+    |=  pax=path
+    ^-  ?
+    ?~  (permission-scry pax)
+      %.n
+    %.y
+  ?.  =((lent checked-paths) perm-length)
+    [~ this]
+  =/  perms  (~(get by relation) group)
+  ::  if relation does not exist, create it and subscribe.
+  ?~  perms
+    =/  group-path  [%group group]
+    :_  this(relation (~(put by relation) group permissions))
+    [ost.bol %peer group-path [our.bol %group-store] group-path]~
+  =.  u.perms  (~(uni in u.perms) permissions)
+  :-  ~
+  this(relation (~(put by relation) group u.perms))
+::
+++  handle-dissociate
+  |=  [group=path permissions=(set path)]
+  ^-  (quip move _this)
+  =/  perms  (~(get by relation) group)
+  ?~  perms
+    [~ this]
+  =.  permissions  (~(del in u.perms) permissions)
+  ?~  permissions
+    :_  this(relation (~(del by relation) group))
+    :~  (group-pull [%group group])
+    ==
+  :-  ~
+  this(relation (~(put by relation) group permissions))
+::
+++  diff-group-update
+  |=  [wir=wire diff=group-update]
+  ^-  (quip move _this)
+  ?-  -.diff
+      %keys
+    [~ this]
+      %bundle
+    [~ this]
+      %path
+    ::  set all permissions paths
+    =/  perms  (~(get by relation) pax.diff)
```

Comment:
The logic here could be simplified using the jug operations (since `relation` is actually a jug):

```
    =/  perms  (~(get ju relation) pax.diff)
    :_  this
    %+  turn  ~(tap in perms)
    |=  =path
    ^-  move
    (permission-poke path [%add path members.diff])
```

Same for the other cases.

Reply:
What is the difference here? Will just return empty set if perms does not exist at the path?

Reply:
Yeah, a `(jug foo bar)` is conceptually a sparse map of all possible `foo` to sets of `bar`s.  So it says that a `foo` not having an entry in the map is the same as having an entry of "empty set".  Often this is a more natural representation if you don't specifically want to handle the case of "have an entry in the map but it's an empty set".  So it tends to reduce branching, as here.

But it's basically an aesthetic decision; choose what seems clearer to you.

Reply:
I'll keep it as is. Thanks for explaining.